src/runtime.rs

```diff
@@ -16,6 +21,34 @@ pub struct TargetRuntime {
     pub cache: Arc<dyn Cache<Key = u64, Value = ConstValue>>,
 }
 
+#[derive(Clone)]
+pub struct TargetRuntimeContext<'a> {
+    pub runtime: &'a TargetRuntime,
+    pub vars: &'a BTreeMap<String, String>,
+    pub headers: HeaderMap, // TODO
+}
+
+impl<'a> PathString for TargetRuntimeContext<'a> {
+    fn path_string<T: AsRef<str>>(&self, path: &[T]) -> Option<Cow<'_, str>> {
+        if path.is_empty() {
+            return None;
+        }
+
+        path.split_first()
+            .and_then(|(head, tail)| match head.as_ref() {
+                "vars" => self.vars.get(tail[0].as_ref()).map(|v| v.into()),
+                "env" => self.runtime.env.get(tail[0].as_ref()).map(|v| v.into()),
+                _ => None,
+            })
+    }
+}
```

Comment:
in main changes with similar structure have been landed here `https://github.com/tailcallhq/tailcall/blob/main/src/config/reader_context.rs` so you can reuse that